advisor/src/main/kotlin/advisors/Osv.kt

```diff
@@ -0,0 +1,199 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.advisor.advisors
+
+import java.net.URI
+import java.time.Instant
+
+import org.ossreviewtoolkit.advisor.AbstractAdviceProviderFactory
+import org.ossreviewtoolkit.advisor.AdviceProvider
+import org.ossreviewtoolkit.clients.osv.Ecosystem
+import org.ossreviewtoolkit.clients.osv.OsvApiClient
+import org.ossreviewtoolkit.clients.osv.OsvService
+import org.ossreviewtoolkit.clients.osv.VulnerabilitiesForPackageRequest
+import org.ossreviewtoolkit.clients.osv.Vulnerability
+import org.ossreviewtoolkit.model.AdvisorCapability
+import org.ossreviewtoolkit.model.AdvisorDetails
+import org.ossreviewtoolkit.model.AdvisorResult
+import org.ossreviewtoolkit.model.AdvisorSummary
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.VulnerabilityReference
+import org.ossreviewtoolkit.model.config.AdvisorConfiguration
+import org.ossreviewtoolkit.utils.common.collectMessages
+import org.ossreviewtoolkit.utils.common.enumSetOf
+import org.ossreviewtoolkit.utils.common.withoutPrefix
+import org.ossreviewtoolkit.utils.ort.OkHttpClientHelper
+import org.ossreviewtoolkit.utils.ort.log
+
+/**
+ * An advice provider that obtains vulnerability from OSV.dev https://osv.dev/.
+ */
+class Osv(name: String, advisorConfiguration: AdvisorConfiguration) : AdviceProvider(name) {
+    class Factory : AbstractAdviceProviderFactory<Osv>("Osv") {
+        override fun create(config: AdvisorConfiguration): Osv =
+            Osv(providerName, config)
+    }
+
+    override val details: AdvisorDetails = AdvisorDetails(providerName, enumSetOf(AdvisorCapability.VULNERABILITIES))
+
+    private val service = OsvService(
+        serverUrl = advisorConfiguration.osv?.serverUrl ?: OsvApiClient.SERVER_URL_PRODUCTION,
+        httpClient = OkHttpClientHelper.buildClient()
+    )
+
+    override suspend fun retrievePackageFindings(packages: List<Package>): Map<Package, List<AdvisorResult>> {
+        val startTime = Instant.now()
+
+        val vulnerabilitiesForPackage = getVulnerabilitiesForPackage(packages)
+
+        log.info { "Found ${vulnerabilitiesForPackage.values.flatten().distinct().size} distinct vulnerabilities." }
+
+        val affectedPackages = vulnerabilitiesForPackage.filter { it.value.isNotEmpty() }.keys
```

Comment:
This makes me wonder why there might be package keys with empty values at all. Can't we simply not add such keys in the first place?

Reply:
Of course we "can", it's however not obvious why we "should". I believe as-is is a tad simpler and also consistent with similar functions, see e.g. https://github.com/oss-review-toolkit/ort/blob/470e2f323b516389bd96a63a81c67b74891a36f8/advisor/src/main/kotlin/advisors/GitHubDefects.kt#L123.